src/transcode.rs

```diff
@@ -0,0 +1,100 @@
+//! to be transcoded to `ExternIO` to be passed through as zome call payload.
+//!
+//! On the way out, the zome call response is `ExternIO` encoded and needs to be converted
+//! to a JSON string.
+
+use crate::HcHttpGatewayResult;
+use base64::{prelude::BASE64_URL_SAFE, Engine};
+use holochain_types::prelude::ExternIO;
+
+/// Function to transcode an incoming base64 encoded payload to Holochain serialized bytes
+/// (type `ExternIO`).
+pub fn base64_json_to_hsb(base64_encoded_payload: &str) -> HcHttpGatewayResult<ExternIO> {
+    let base64_decoded_payload = BASE64_URL_SAFE.decode(base64_encoded_payload)?;
+    let json_payload = serde_json::from_slice::<serde_json::Value>(&base64_decoded_payload)?;
+    let msgpack_encoded_payload = ExternIO::encode(json_payload)?;
+    Ok(msgpack_encoded_payload)
+}
+
+/// Function to transcode a zome call response encoded as Holochain serialized bytes (type `ExternIO`)
+/// to a JSON string.
+pub fn hsb_to_json(hsb_encoded_response: &ExternIO) -> anyhow::Result<String> {
+    let json_value = hsb_encoded_response.decode::<serde_json::Value>()?;
+    Ok(json_value.to_string())
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::{
+        transcode::{base64_json_to_hsb, hsb_to_json},
+        HcHttpGatewayError,
+    };
+    use base64::{prelude::BASE64_URL_SAFE, Engine};
+    use holochain_types::prelude::ExternIO;
+    use serde::{Deserialize, Serialize};
+
+    #[test]
+    fn happy_base64_json_to_hsb() {
+        #[derive(Debug, PartialEq, Deserialize, Serialize)]
+        struct ZomeCallPayload {
+            field: bool,
+        }
+
+        // Create a payload, serialize to JSON and base64 encode.
+        let payload = ZomeCallPayload { field: false };
+        let json_payload = serde_json::to_string(&payload).unwrap();
+        let base64_encoded_payload = BASE64_URL_SAFE.encode(json_payload);
+
+        let hsb_encoded_payload = base64_json_to_hsb(&base64_encoded_payload).unwrap();
+
+        // Deserializing the serialized bytes to the original struct should succeed.
+        let decoded_payload = hsb_encoded_payload.decode::<ZomeCallPayload>().unwrap();
+        assert_eq!(decoded_payload, payload);
+    }
+
+    #[test]
+    fn plain_json_to_hsb_fails() {
+        #[derive(Debug, PartialEq, Deserialize, Serialize)]
+        struct ZomeCallPayload {
+            field: bool,
+        }
+
+        let payload = ZomeCallPayload { field: false };
+        let json_payload = serde_json::to_string(&payload).unwrap();
+
+        let result = base64_json_to_hsb(&json_payload);
+        assert!(matches!(
+            result,
+            Err(HcHttpGatewayError::Base64DecodeError(_))
+        ));
+    }
+
+    #[test]
+    fn invalid_json_to_hsb_fails() {
+        let base64_encoded_payload = BASE64_URL_SAFE.encode("invalid");
+
+        let result = base64_json_to_hsb(&base64_encoded_payload);
+        assert!(matches!(result, Err(HcHttpGatewayError::InvalidJSON(_))));
+    }
+
```

Comment:
I couldn't come up with a (reasonable) test to produce a failing `ExternIO` encoding. [`SerializedBytesError`](https://github.com/holochain/holochain-serialization/blob/main/crates/holochain_serialized_bytes/src/lib.rs#L64-L71) also considers it very unlikely.